read_utils.py

```diff
@@ -1208,6 +1208,12 @@ def align_and_fix(
         else:
             aligner_options = '' # use defaults
 
+    inBamFiltered = mkstempfname('.input-filtered.bam')
+    samtools.filter_to_mapped_reads(inBam, inBamFiltered, allow_unmapped=True)
+    filtered_count = samtools.count(inBamFiltered)
+    if filtered_count==0:
+        log.warning("zero reads after input filtering")
+
```

Comment:
Hm.. why are we calling the `filter_to_mapped_reads` twice (before and after alignment) and not just once at the end (post-alignment, as originally)?

In the particular problem you had with your dataset, I think this first filter wouldn't actually do much of anything: the unmapped reads were all paired and fine -- the problem was with the alignments, and it wouldn't be visible until after the aligner did its thing.

Reply:
The input filtering is mostly as a guard to remove singletons from problematic or already aligned-and-filtered input (quicker than RevertSam SANITIZE=True). It was useful when re-running the aligned reads from the recent dataset through the pipeline again since some were singletons. We can probably omit it if we know the input will be unaligned and consistent, but it doesn't add that much time as safety check. minimap2 will complain if it sees pairs with missing mates, but bwa seems to align them anyway and mark them as single-end, so with bwa singletons that are not filtered on the input side then pass the filter on the output side…